File: working_hours/application/user_service.py
"""User management use cases: auth, password ops, admin CRUD, employee CRUD.

All password hashing lives here — nowhere else imports werkzeug.security directly.
"""
from __future__ import annotations

from werkzeug.security import check_password_hash, generate_password_hash

from ..domain.validators import is_username_taken, validate_password
from ..infrastructure.settings import _find_user, _load_settings, _save_settings


# ---------------------------------------------------------------------------
# Password primitives
# ---------------------------------------------------------------------------

def hash_password(raw: str) -> str:
    return generate_password_hash(raw)


def verify_password(raw: str, hashed: str) -> bool:
    return check_password_hash(hashed, raw)


# ---------------------------------------------------------------------------
# Authentication
# ---------------------------------------------------------------------------

def find_user(username: str) -> tuple[dict | None, str | None, bool]:
    """Delegate to settings layer; returns (user_dict, emp_id, is_admin)."""
    return _find_user(username)


def authenticate(
    username: str, password: str
) -> tuple[dict | None, str | None, bool]:
    """Return (user, emp_id, is_admin) if credentials are valid and account enabled.

    Returns (None, None, False) on any failure.
    """
    user, emp_id, is_admin = _find_user(username)
    if not user:
        return None, None, False
    if not verify_password(password, user.get("password_hash", "")):
        return None, None, False
    if emp_id is not None and not user.get("enabled", True):
        return None, None, False
    return user, emp_id, is_admin


# ---------------------------------------------------------------------------
# First-run setup
# ---------------------------------------------------------------------------

def needs_setup() -> bool:
    return not _load_settings().get("admin_users")


def create_initial_admin(username: str, password: str, confirm: str) -> str | None:
    """Create the very first admin account. Returns an error string or None."""
    if not username:
        return "Username is required."
    if not password:
        return "Password is required."
    err = validate_password(password, confirm)
    if err:
        return err
    settings = _load_settings()
    settings.setdefault("admin_users", {})[username] = {
        "password_hash": hash_password(password),
    }
    _save_settings(settings)
    return None


# ---------------------------------------------------------------------------
# Self-service password change
# ---------------------------------------------------------------------------

_MISSING_FIELDS = "MISSING_FIELDS"
_NOT_FOUND = "NOT_FOUND"
_WRONG_PASSWORD = "WRONG_PASSWORD"


def change_password(username: str, current_pw: str, new_pw: str) -> str | None:
    """Change a user's own password. Returns a sentinel string or None on success."""
    if not current_pw or not new_pw:
        return _MISSING_FIELDS
    user, emp_id, is_admin = _find_user(username)
    if not user:
        return _NOT_FOUND
    if not verify_password(current_pw, user.get("password_hash", "")):
        return _WRONG_PASSWORD
    settings = _load_settings()
    new_hash = hash_password(new_pw)
    if is_admin:
        settings["admin_users"][username]["password_hash"] = new_hash
    else:
        settings["employees"][emp_id]["password_hash"] = new_hash
    _save_settings(settings)
    return None


# ---------------------------------------------------------------------------
# Admin account CRUD
# ---------------------------------------------------------------------------

def list_admins() -> list[dict]:
    return [{"username": u} for u in _load_settings().get("admin_users", {})]


def create_admin(username: str, password: str) -> str | None:
    """Create an admin account. Returns an error sentinel or None."""
    if not username or not password:
        return _MISSING_FIELDS
    settings = _load_settings()
    if is_username_taken(username, settings):
        return "USERNAME_TAKEN"
    settings.setdefault("admin_users", {})[username] = {
        "password_hash": hash_password(password),
    }
    _save_settings(settings)
    return None


def update_admin_password(username: str, new_password: str) -> str | None:
    """Update an admin's password. Returns an error sentinel or None."""
    settings = _load_settings()
    if username not in settings.get("admin_users", {}):
        return _NOT_FOUND
    if new_password:
        settings["admin_users"][username]["password_hash"] = hash_password(new_password)
    _save_settings(settings)
    return None


def delete_admin(username: str, current_username: str | None) -> str | None:
    """Delete an admin account. Returns an error sentinel or None."""
    if username == "admin":
        return "PROTECTED"
    if username == current_username:
        return "SELF_DELETE"
    settings = _load_settings()
    if username not in settings.get("admin_users", {}):
        return _NOT_FOUND
    del settings["admin_users"][username]
    _save_settings(settings)
    return None


# ---------------------------------------------------------------------------
# Employee CRUD
# ---------------------------------------------------------------------------

def list_employees(events: list) -> list[dict]:
    """Build employee list merging raw event data with settings."""
    known: dict[str, str] = {}
    for e in events:
        known[str(e.emp_id)] = e.name
    settings = _load_settings()
    employees = settings.get("employees", {})
    result = []
    for emp_id in sorted(known.keys(), key=lambda x: x.zfill(20) if x.isdigit() else x):
        emp = employees.get(emp_id, {})
        result.append({
            "emp_id": emp_id,
            "raw_name": known[emp_id],
            "alias": emp.get("alias", ""),
            "full_name": emp.get("full_name", ""),
            "username": emp.get("username", ""),
            "has_password": bool(emp.get("password_hash", "")),
            "enabled": emp.get("enabled", True),
        })
    return result


def update_employee(emp_id: str, data: dict) -> str | None:
    """Update employee record fields. Returns an error sentinel or None."""
    settings = _load_settings()
    emp = settings.setdefault("employees", {}).setdefault(emp_id, {})
    new_username = data.get("username", emp.get("username", "")).strip()
    if new_username and new_username != emp.get("username", ""):
        if is_username_taken(new_username, settings, exclude_emp_id=emp_id):
            return "USERNAME_TAKEN"
    if "alias" in data:
        emp["alias"] = data["alias"].strip()
    if "full_name" in data:
        emp["full_name"] = data["full_name"].strip()
    emp["username"] = new_username
    if "enabled" in data:
        emp["enabled"] = bool(data["enabled"])
    if data.get("password"):
        emp["password_hash"] = hash_password(data["password"])
    _save_settings(settings)
    return None


def get_profiles() -> dict:
    """Return {emp_id: {alias, full_name}} for all employees."""
    employees = _load_settings().get("employees", {})
    return {
        emp_id: {"alias": emp.get("alias", ""), "full_name": emp.get("full_name", "")}
        for emp_id, emp in employees.items()
    }

File: packages/server/src/application/userService.ts
import { pbkdf2Sync, timingSafeEqual } from 'crypto'
import bcrypt from 'bcryptjs'
import type { ClockEvent } from '../domain/models.js'
import { findUser, loadSettings, saveSettings } from '../infrastructure/settings.js'

// ---------------------------------------------------------------------------
// Password hashing — supports existing werkzeug PBKDF2 hashes + new bcrypt
// ---------------------------------------------------------------------------

function verifyWerkzeugHash(password: string, stored: string): boolean {
  const parts = stored.split('$')
  if (parts.length !== 3) return false
  const [methodStr, salt, expected] = parts
  const methodParts = methodStr.split(':')
  if (methodParts[0] !== 'pbkdf2' || methodParts.length < 3) return false
  const digestAlgo = methodParts[1]
  const iterations = parseInt(methodParts[2])
  if (isNaN(iterations)) return false
  const derived = pbkdf2Sync(password, salt, iterations, 32, digestAlgo)
  const derivedBuf = Buffer.from(derived.toString('hex'))
  const expectedBuf = Buffer.from(expected)
  if (derivedBuf.length !== expectedBuf.length) return false
  return timingSafeEqual(derivedBuf, expectedBuf)
}

export function hashPassword(raw: string): string {
  return bcrypt.hashSync(raw, 12)
}

export function verifyPassword(raw: string, stored: string): boolean {
  if (stored.startsWith('pbkdf2:')) return verifyWerkzeugHash(raw, stored)
  return bcrypt.compareSync(raw, stored)
}

// ---------------------------------------------------------------------------
// Authentication
// ---------------------------------------------------------------------------

export interface AuthResult {
  ok: boolean
  empId: string | null
  isAdmin: boolean
}

export function authenticate(username: string, password: string): AuthResult {
  const { user, empId, isAdmin } = findUser(username)
  if (!user) return { ok: false, empId: null, isAdmin: false }
  if (!verifyPassword(password, (user as any).password_hash ?? '')) return { ok: false, empId: null, isAdmin: false }
  if (empId !== null && (user as any).enabled === false) return { ok: false, empId: null, isAdmin: false }
  return { ok: true, empId, isAdmin }
}

// ---------------------------------------------------------------------------
// Setup
// ---------------------------------------------------------------------------

export function needsSetup(): boolean {
  return !Object.keys(loadSettings().admin_users ?? {}).length
}

export function createInitialAdmin(username: string, password: string, confirm: string): string | null {
  if (!username) return 'Username is required.'
  if (!password) return 'Password is required.'
  if (password.length < 8) return 'Password must be at least 8 characters.'
  if (password !== confirm) return 'Passwords do not match.'
  const s = loadSettings()
  s.admin_users ??= {}
  s.admin_users[username] = { password_hash: hashPassword(password) }
  saveSettings(s)
  return null
}

// ---------------------------------------------------------------------------
// Self-service password change
// ---------------------------------------------------------------------------

export const ERR_MISSING = 'MISSING_FIELDS'
export const ERR_NOT_FOUND = 'NOT_FOUND'
export const ERR_WRONG_PW = 'WRONG_PASSWORD'
export const ERR_USERNAME_TAKEN = 'USERNAME_TAKEN'

export function changePassword(username: string, currentPw: string, newPw: string): string | null {
  if (!currentPw || !newPw) return ERR_MISSING
  const { user, empId, isAdmin } = findUser(username)
  if (!user) return ERR_NOT_FOUND
  if (!verifyPassword(currentPw, (user as any).password_hash ?? '')) return ERR_WRONG_PW
  const s = loadSettings()
  const newHash = hashPassword(newPw)
  if (isAdmin) {
    s.admin_users[username].password_hash = newHash
  } else if (empId) {
    s.employees[empId].password_hash = newHash
  }
  saveSettings(s)
  return null
}

// ---------------------------------------------------------------------------
// Admin CRUD
// ---------------------------------------------------------------------------

export function listAdmins(): { username: string }[] {
  return Object.keys(loadSettings().admin_users ?? {}).map(u => ({ username: u }))
}

export function createAdmin(username: string, password: string): string | null {
  if (!username || !password) return ERR_MISSING
  const s = loadSettings()
  if (isUsernameTaken(username, s)) return 'USERNAME_TAKEN'
  s.admin_users ??= {}
  s.admin_users[username] = { password_hash: hashPassword(password) }
  saveSettings(s)
  return null
}

export function updateAdminPassword(username: string, newPassword: string): string | null {
  const s = loadSettings()
  if (!s.admin_users?.[username]) return ERR_NOT_FOUND
  if (newPassword) s.admin_users[username].password_hash = hashPassword(newPassword)
  saveSettings(s)
  return null
}

export function deleteAdmin(username: string, currentUsername: string | null): string | null {
  if (username === 'admin') return 'PROTECTED'
  if (username === currentUsername) return 'SELF_DELETE'
  const s = loadSettings()
  if (!s.admin_users?.[username]) return ERR_NOT_FOUND
  delete s.admin_users[username]
  saveSettings(s)
  return null
}

// ---------------------------------------------------------------------------
// Employee CRUD
// ---------------------------------------------------------------------------

export function listEmployees(events: ClockEvent[]): object[] {
  const known = new Map<string, string>()
  for (const e of events) known.set(e.empId, e.name)
  const s = loadSettings()
  const employees = s.employees ?? {}
  return [...known.keys()]
    .sort((a, b) => {
      const an = a.match(/^\d+$/) ? a.padStart(20, '0') : a
      const bn = b.match(/^\d+$/) ? b.padStart(20, '0') : b
      return an.localeCompare(bn)
    })
    .map(empId => {
      const emp = employees[empId] ?? {}
      return {
        emp_id: empId,
        raw_name: known.get(empId),
        alias: emp.alias ?? '',
        full_name: emp.full_name ?? '',
        email: emp.email ?? '',
        username: emp.username ?? '',
        has_password: Boolean(emp.password_hash),
        enabled: emp.enabled !== false,
      }
    })
}

export function updateEmployee(empId: string, data: Record<string, unknown>): string | null {
  const s = loadSettings()
  s.employees ??= {}
  s.employees[empId] ??= { alias: '', full_name: '', username: '', password_hash: '', is_admin: false, enabled: true }
  const emp = s.employees[empId]
  const newUsername = ((data.username as string | undefined) ?? emp.username ?? '').trim()
  if (newUsername && newUsername !== emp.username) {
    if (isUsernameTaken(newUsername, s, empId)) return 'USERNAME_TAKEN'
  }
  if ('alias' in data) emp.alias = (data.alias as string).trim()
  if ('full_name' in data) emp.full_name = (data.full_name as string).trim()
  if ('email' in data) emp.email = ((data.email as string) ?? '').trim()
  emp.username = newUsername
  if ('enabled' in data) emp.enabled = Boolean(data.enabled)
  if (data.password) emp.password_hash = hashPassword(data.password as string)
  saveSettings(s)
  return null
}

export function updateProfile(
  username: string,
  data: { full_name?: string; email?: string; username?: string; current_password?: string; new_password?: string }
): string | null {
  const { user, empId } = findUser(username)
  if (!user || !empId) return ERR_NOT_FOUND
  const s = loadSettings()
  const emp = s.employees[empId]
  const newUsername = (data.username ?? emp.username ?? '').trim()
  if (newUsername && newUsername !== emp.username) {
    if (isUsernameTaken(newUsername, s, empId)) return ERR_USERNAME_TAKEN
  }
  if (data.current_password && data.new_password) {
    if (!verifyPassword(data.current_password, emp.password_hash ?? '')) return ERR_WRONG_PW
    emp.password_hash = hashPassword(data.new_password)
  }
  if ('full_name' in data) emp.full_name = (data.full_name ?? '').trim()
  if ('email' in data) emp.email = (data.email ?? '').trim()
  emp.username = newUsername || emp.username
  saveSettings(s)
  return null
}

export function getUserConfig(empId: string): { full_name: string; email: string } {
  const emp = loadSettings().employees[empId]
  return { full_name: emp?.full_name ?? '', email: emp?.email ?? '' }
}

export function getProfiles(): Record<string, { alias: string; full_name: string }> {
  const employees = loadSettings().employees ?? {}
  return Object.fromEntries(
    Object.entries(employees).map(([id, emp]) => [id, { alias: emp.alias ?? '', full_name: emp.full_name ?? '' }])
  )
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isUsernameTaken(username: string, s: ReturnType<typeof loadSettings>, excludeEmpId?: string): boolean {
  if (s.admin_users?.[username]) return true
  for (const [id, emp] of Object.entries(s.employees ?? {})) {
    if (id === excludeEmpId) continue
    if (emp.username === username) return true
  }
  return false
}
